File: Codigo/Nivel 1/Nivel 2/Nivel 3/Nivel 4/Ejemplos/activeMQ_python_ejemplo/MTISSubscribeAsync.py
import stomp
import threading

class AsyncListener(stomp.ConnectionListener):
    def __init__(self, event):
        self.event = event

    def on_message(self, frame):
        print(f"Received message: {frame.body}")
        self.event.set()

def main():
    # Utilizamos un threading.Event para esperar un mensaje.
    event = threading.Event()
    
    conn = stomp.Connection12(host_and_ports=[("localhost", 61613)])
    listener = AsyncListener(event)
    conn.set_listener('', listener)
    conn.connect(wait=True)
    
    destination = "/topic/historiales"
    conn.subscribe(destination=destination, id=1, ack='auto')
    print("Waiting for asynchronous message...")
    
    # Espera hasta recibir el mensaje (timeout de 10 segundos)
    if event.wait(timeout=10000):
        print("Message received, disconnecting.")
    else:
        print("Timeout waiting for message.")
    
    conn.disconnect()

if __name__ == "__main__":
    main()

File: Codigo/Nivel 1/Nivel 2/Nivel 3/Nivel 4/Ejemplos/activeMQ_php_ejemplo/MTISSubscribeAsync.php
<?php
// MTISSubscribeAsync.php

require_once __DIR__ . '/vendor/autoload.php';

use Stomp\Client;
use Stomp\StatefulStomp;

$url = 'tcp://localhost:61613';
$topic = '/topic/historiales';

// Creamos el cliente STOMP
$client = new Client($url);
$client->setLogin('mtis', 'mtis');

$stomp = new StatefulStomp($client);

// Nos suscribimos al tema (modo AUTO_ACKNOWLEDGE por defecto)
$stomp->subscribe($topic);

echo "Waiting for messages asynchronously..." . PHP_EOL;

// Bucle infinito para esperar mensajes de forma continua
while (true) {
    // read() es bloqueante y esperará hasta recibir un mensaje
    $frame = $stomp->read();
    if ($frame) {
        echo "Received message: " . $frame->body . PHP_EOL;
        // No es necesario llamar a ack() en modo AUTO_ACKNOWLEDGE
    }
    sleep(1);
}
